File: 프로그래머스/2/68645. 삼각 달팽이/삼각 달팽이.java
class Solution {
    public int[] solution(int n) {
        
        int len = n*(n+1)/2;
        int[] answer = new int[len];
        
        int[][] map = new int[n][n];
        
        // 순서: 아래 오른쪽 위
        int[] dx = {1, 0, -1};
        int[] dy = {0, 1, -1};
        
        int num = 1;
        int dir = 0;
        int nx = 0, ny = 0,  x = 0, y = 0;
        
        for(int i=0; i<len; ++i){
            map[x][y] = num;
            ++num;
            
            nx = x+dx[dir];
            ny = y+dy[dir];
            
            // 방향 전환 여부 판단
            if(nx < 0 || nx >= n || ny < 0 || ny >= n || map[nx][ny] != 0){
                dir = (dir+1)%3;
                
                nx = x+dx[dir];
                ny = y+dy[dir];
            }
            
            x = nx;
            y = ny;
        }
        
        int idx = 0;
        for(int i=0; i<n; ++i){
            for(int j=0; j<=i; ++j){
                answer[idx] = map[i][j];
                ++idx;
            }
        }
        
        return answer;
    }
}

File: 프로그래머스/2/68645. 삼각 달팽이/삼각 달팽이.py
def solution(n):
    l = (n*(n+1))//2
    answer = [0 for _ in range(l)]
    
    map = [[0 for _ in range(n)] for _ in range(n)]
    
    dir = 0
    x = y = nx = ny = 0
    
    dx = [1, 0, -1]
    dy = [0, 1, -1]
    
    for num in range(1, l+1):
        map[x][y] = num;
        
        nx = x+dx[dir]
        ny = y+dy[dir]
        
        if nx < 0 or nx >= n or ny < 0 or ny >= n or map[nx][ny] != 0:
            dir = (dir+1)%3
            nx = x+dx[dir]
            ny = y+dy[dir]
            
        x = nx
        y = ny
           
    idx = 0   
    for i in range(n):
        for j in range(i+1):
            answer[idx] = map[i][j]
            idx += 1
    
    return answer